ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java

```diff
@@ -0,0 +1,291 @@
+     */
+    public void createInteraction(
+        String conversationId,
+        String input,
+        String promptTemplate,
+        String response,
+        String origin,
+        Map<String, String> additionalInfo,
+        String parentIntId,
+        Integer traceNum,
+        ActionListener<CreateInteractionResponse> actionListener
+    ) {
+        Preconditions.checkNotNull(conversationId);
+        Preconditions.checkNotNull(input);
+        Preconditions.checkNotNull(response);
+        // additionalInfo cannot be null as flat object
+        additionalInfo = (additionalInfo == null) ? new HashMap<>() : additionalInfo;
+        try {
+            client
+                .execute(
+                    CreateInteractionAction.INSTANCE,
+                    new CreateInteractionRequest(
+                        conversationId,
+                        input,
+                        promptTemplate,
+                        response,
+                        origin,
+                        additionalInfo,
+                        parentIntId,
+                        traceNum
+                    ),
+                    actionListener
+                );
+        } catch (RuntimeException exception) {
+            actionListener.onFailure(exception);
+        }
+    }
+
+    /**
+     * Get the interactions associate with this conversation, sorted by recency
+     * @param conversationId the conversation whose interactions to get
+     * @param lastNInteraction Return how many interactions
+     * @param actionListener get all the final interactions that are not traces
+     */
+    public void getFinalInteractions(String conversationId, int lastNInteraction, ActionListener<List<Interaction>> actionListener) {
+        Preconditions.checkArgument(lastNInteraction > 0, "lastN must be at least 1.");
+        log.info("Getting Interactions, conversationId {}, lastN {}", conversationId, lastNInteraction);
+
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().newStoredContext(true)) {
+            if (!clusterService.state().metadata().hasIndex(indexName)) {
+                actionListener.onResponse(List.of());
+                return;
+            }
+            ActionListener<Boolean> accessListener = ActionListener.wrap(access -> {
+                if (access) {
+                    innerGetFinalInteractions(conversationId, lastNInteraction, actionListener);
+                } else {
+                    String userstr = client
+                        .threadPool()
+                        .getThreadContext()
+                        .getTransient(ConfigConstants.OPENSEARCH_SECURITY_USER_INFO_THREAD_CONTEXT);
+                    String user = User.parse(userstr) == null ? ActionConstants.DEFAULT_USERNAME_FOR_ERRORS : User.parse(userstr).getName();
+                    throw new OpenSearchSecurityException("User [" + user + "] does not have access to conversation " + conversationId);
```

Comment:
what do you think about this kind of exception?

```
new OpenSearchStatusException(
                                    "User [" + user + "] does not have access to conversation " + conversationId,
                                    RestStatus.FORBIDDEN
                                )
                                ```

Reply:
I'll remove this class due to a new one in Xun's pr.